File: exp1/src/main/java/com/example/util/dataProcessing.java
package com.example.util;

import com.example.entity.Student;

import java.util.ArrayList;
import java.util.HashMap;

public class dataProcessing {
    public static double GetAverage(ArrayList<Double> grades) {   //传入一个grades的ArrayList求均值
        double sum = 0;   //sum，用于保存列表所有元素的和，初始化为0
        for (int i = 0; i < grades.size(); i++) {  //对列表中的所有元素进行求和
            sum += grades.get(i);
        }
        return sum / grades.size();   //return average
    }

    public static double GetCovariance(ArrayList<Double> grades) {  //传入一个grades的ArrayList求方差
        double avg = dataProcessing.GetAverage(grades);     //调用实现的方法求平均值
        double sum = 0;  //sum
        for (int i = 0; i < grades.size(); i++) {
            sum += (grades.get(i) - avg) * (grades.get(i) - avg);   //sum = (Xi-avg)^2
        }
        return sum / (grades.size() - 1);   //return cov
    }

    public static ArrayList<Double> GetZScore(ArrayList<Double> grades) {  //传入一个grades的ArrayList，进行Zscore化
        double avg = dataProcessing.GetAverage(grades);   //调用平均值函数求均值
        double s = Math.sqrt(dataProcessing.GetCovariance(grades));  //求出方法
        ArrayList<Double> zscore = new ArrayList<>();
        for (int i = 0; i < grades.size(); i++) {
            zscore.add((grades.get(i) - avg) / s);
        }
        return zscore;   //return zscore
    }

    public static Double GetCorrelation(ArrayList<Double> grades1, ArrayList<Double> grades2) { //求grades1和grades2的相关系数
        ArrayList<Double> A = dataProcessing.GetZScore(grades1);
        ArrayList<Double> B = dataProcessing.GetZScore(grades2);
        double sum = 0;
        for (int i = 0; i < grades1.size(); i++) {
            sum += A.get(i) * B.get(i);
        }
        return sum;    //return Corr
    }

    public static Double TransformConstitution(String constitute){   //将体测成绩转换成百分制成绩
            switch (constitute){
                case "excellent": return 100.0;   //excellent -->100
                case "good": return 80.0;   //good --> 80
                case "general":return 60.0;  //general --> 60
                case "bad":return 40.0;  //bad --> 40
                default:return 0.0;  //default 0
            }
    }

    public static void TransformMtoCm(ArrayList<Student> students){  //转换m to cm
        for (Student stu:     //foreach stu in students
             students) {
            stu.setHeight(stu.getHeight()*100);
        }

    }

    public static void TransformCmToM(ArrayList<Student> students){  //转换cm to m
        for (Student stu:     //foreach stu in students
             students) {
            stu.setHeight(stu.getHeight()/100);
        }
    }

    public static ArrayList<Student> MergeCsvAndTxt(String filePath_csv,String filePath_txt){  //合并Csv和Txt文件
        ArrayList<Student> list = ReadFile.ReadCsv(filePath_csv);   //读取Csv的内容
        ArrayList<Student> list_txt = ReadFile.ReadCsv(filePath_txt);  //读取txt
        dataProcessing.TransformMtoCm(list_txt);   //将Txt文件中的内容的身高转换为cm
        list.addAll(list_txt);  //添加Txt的内容
        return list;    //返回学生列表
    }

    public static ArrayList<Student> DuplicateRemoval(ArrayList<Student> students){  //传入一个Student的ArrayList，根据Name去重
        HashMap<String,Student> hashMap = new HashMap<>();
        ArrayList<Student> stuList = new ArrayList<>();
        for (int i =0;i<students.size();i++){
            String key = students.get(i).getName();   //key
            Student value = students.get(i);
            hashMap.put(key,value);
        }

        for (String stuName:
             hashMap.keySet()) {
            stuList.add(hashMap.get(stuName));
        }
        return stuList;
    }

    /**
     * 合并并去重
     * @param filePath_csv
     * @param filePath_txt
     * @return
     */
    public static ArrayList<Student> MergeDuplicateRemoval_CsvAndTxt(String filePath_csv,String filePath_txt){
        ArrayList<Student> students = dataProcessing.MergeCsvAndTxt(filePath_csv,filePath_txt);
        return dataProcessing.DuplicateRemoval(students);
    }


}

File: exp1/python/dataProcessing.py
import numpy as np

def GetAverage(x):         #求均值
    sum = 0
    for i in x:   #对数组中的元素进行求和
        sum += i
    return sum/len(x)  #return 和除以长度（均值）

def GetCovariance(x):      #求方差
    x1 = GetAverage(x)  #求出平均值，保存到x中
    sum = 0
    for i in x:
        sum += (i-x1)**2  #sum =(x-x1)*(x-x1)
    return sum/(len(x) - 1)   #return cov

def GetZScore(x):         #ZScore化
    avg = GetAverage(x)  #cov
    s = np.sqrt(GetCovariance(x))   #s
    y = []
    for i in range(len(x)):
        if s != 0:
            x[i] = (x[i]-avg)/s
    return x   #return x by format

def GetCorrelation(x,y):    #求相关系数
    A = GetZScore(x)   #get x format by zscore  
    B = GetZScore(y)   #get x format by zscore  
    sum = 0
    for i in range(len(x)):     # x,y长度一样
        sum += A[i] * B[i]
    return sum   #return corr

def GetZscoreArray(x):   #求Zscore数组
    Matrix = []
    for i in x:
        Matrix.append(GetZScore(i))
    return Matrix

def GetMax(x):        #获取列表中最大的数
    max = x[0]
    for i in range(len(x)):   #遍历x列表，求出最大值max
        if max < x[i]:   #如果数组中找到比它更大的数，用数组中对应的元素赋值给max
            max = x[i]
    return max  #return max

def GetNumberOfMax(num,x):   #获取列表最大的几个数
    max_of_number =[]    #初始化最大数组为空列表
    for i in range(num):   #根据给定的num进行循环
        max = GetMax(x)   #获取列表的最大值
        max_of_number.append(max)  #向max_of_number中添加当前列表中最大的值
        x.remove(max)   #列表中移除掉最大值
    return max_of_number   #return max of number